plugins/storage/volume/flasharray/src/main/java/org/apache/cloudstack/storage/datastore/adapter/flasharray/FlashArrayAdapter.java

```diff
@@ -0,0 +1,1084 @@
+        ProviderVolume vol = getVolume(dataObject.getExternalName());
+        Long usedBytes = vol.getUsedBytes();
+        Long allocatedSizeInBytes = vol.getAllocatedSizeInBytes();
+        if (usedBytes == null || allocatedSizeInBytes == null) {
+            return null;
+        }
+        ProviderVolumeStats stats = new ProviderVolumeStats();
+        stats.setAllocatedInBytes(allocatedSizeInBytes);
+        stats.setActualUsedInBytes(usedBytes);
+        return stats;
+    }
+
+    @Override
+    public boolean canAccessHost(ProviderAdapterContext context, String hostname) {
+        if (hostname == null) {
+            throw new RuntimeException("Unable to validate host access because a hostname was not provided");
+        }
+
+        List<String> members = getHostgroupMembers(hostgroup);
+
+        // check for fqdn and shortname combinations.  this assumes there is at least a shortname match in both the storage array and cloudstack
+        // hostname configuration
+        String shortname;
+        if (hostname.indexOf('.') > 0) {
+            shortname = hostname.substring(0, (hostname.indexOf('.')));
+        } else {
+            shortname = hostname;
+        }
+
+        for (String member : members) {
+            // exact match (short or long names)
+            if (member.equals(hostname)) {
+                return true;
+            }
+
+            // primera has short name and cloudstack had long name
+            if (member.equals(shortname)) {
+                return true;
+            }
+
+            // member has long name but cloudstack had shortname
+            if (member.indexOf('.') > 0) {
+                if (member.substring(0, (member.indexOf('.'))).equals(shortname)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private String getAccessToken() {
+        refreshSession(false);
+        return accessToken;
+    }
+
+    private synchronized void refreshSession(boolean force) {
+        try {
+            if (force || keyExpiration < System.currentTimeMillis()) {
+                // close client to force connection reset on appliance -- not doing this can
+                // result in NotAuthorized error...guessing
+                _client.close();
+                ;
+                _client = null;
+                login();
+                keyExpiration = System.currentTimeMillis() + keyTtl;
+            }
+        } catch (Exception e) {
+            // retry frequently but not every request to avoid DDOS on storage API
+            logger.warn("Failed to refresh FlashArray API key for " + username + "@" + url + ", will retry in 5 seconds",
+                    e);
+            keyExpiration = System.currentTimeMillis() + (5 * 1000);
+        }
+    }
+
+    /**
+     * Login to the array and get an access token
+     */
+    private void login() {
```

Comment:
long method. can it be cut into smaller methods?

Reply:
Updated in upcoming commit.